core/src/main/java/org/mule/runtime/core/internal/routing/ChoiceRouter.java

```diff
@@ -6,51 +6,190 @@
+    }
+  }
+
+  @Override
+  public void start() throws MuleException {
+    for (ProcessorRoute route : routes) {
+      route.start();
+    }
+
+    started.set(true);
+  }
+
+  @Override
+  public void stop() throws MuleException {
+    for (ProcessorRoute route : routes) {
+      route.stop();
+    }
+
+    started.set(false);
+  }
+
+  @Override
+  public void dispose() {
+    for (ProcessorRoute route : routes) {
+      route.dispose();
+    }
+  }
+
+  public void addRoute(final String expression, final Processor processor) {
+    routes.add(new ProcessorExpressionRoute(expression, processor));
+  }
+
+  public void setDefaultRoute(final Processor processor) {
+    defaultProcessor = processor;
+  }
+
+  @Override
+  public CoreEvent process(CoreEvent event) throws MuleException {
+    return processToApply(event, this);
+  }
+
+  @Override
+  public Publisher<CoreEvent> apply(Publisher<CoreEvent> publisher) {
+    return Flux.merge(new SinkRouter(publisher, routes).collectPublishers());
+  }
+
+  public void updateStatistics(Processor processor) {
+    if (getRouterStatistics() != null && getRouterStatistics().isEnabled()) {
+      getRouterStatistics().incrementRoutedMessage(processor);
+    }
+  }
+
+  public RouterStatistics getRouterStatistics() {
+    return routerStatistics;
+  }
+
+  @Override
+  public void setRouterStatistics(RouterStatistics routerStatistics) {
+    this.routerStatistics = routerStatistics;
+  }
+
+  @Override
+  public String toString() {
+    return format("%s [flow=%s, started=%s]", getClass().getSimpleName(), getLocation().getRootContainerName(), started);
+  }
+
+  /**
+   * Router that generates executable routes for each one configured and handles the logic of deciding which one will be executed
+   * when an event arrives. It also provides access to the routes publishers so they can be merged together.
+   */
+  private class SinkRouter {
+
+    private final Flux<CoreEvent> router;
+    private final List<ExecutableRoute> routes;
+
+    SinkRouter(Publisher<CoreEvent> publisher, List<ProcessorRoute> routes) {
+      this.routes = routes.stream().map(ProcessorRoute::toExecutableRoute).collect(toList());
+      router = from(publisher)
+          .doOnNext(checkedConsumer(this::route))
+          .doOnComplete(() -> this.routes.stream()
+              .forEach(executableRoute -> executableRoute.complete()));
+    }
+
+    /**
+     * Decides which route should execute for an incoming event and executes it, purposely separating those actions so that a
+     * single {@link ExpressionManagerSession} can be used for all routes decision process.
+     *
+     * @param event the incoming event
+     */
+    private void route(CoreEvent event) {
+      ExecutableRoute selectedRoute;
+      try (ExpressionManagerSession session = expressionManager.openSession(getLocation(), event, NULL_BINDING_CONTEXT)) {
+        selectedRoute = routes.stream().filter(route -> route.shouldExecute(session)).findFirst().get();
```

Comment:
add a comment explaining that there will always be an element for the defualt route at the end of the list